lexer/lexer.go

```diff
@@ -11,18 +11,98 @@ type Lexer struct {
 	position     int  // current position in input (points to current char)
 	readPosition int  // current reading position in input (after current char)
 	ch           byte // current char under examination
+	// map of input line boundaries used by LinePosition() for error location
+	lineMap [][2]int // array of [begin, end] pairs: [[0,12], [13,22], [23,33] ... ]
 }
 
 func New(input string) *Lexer {
 	l := &Lexer{input: input}
+	// map the input line boundaries for CurrentLine()
+	l.buildLineMap()
+	// read the first char
 	l.readChar()
 	return l
 }
 
+// buildLineMap creates map of input line boundaries used by LinePosition() for error location
+func (l *Lexer) buildLineMap() {
+	begin := 0
+	idx := 0
+	for i, ch := range l.input {
+		idx = i
+		if ch == '\n' {
+			l.lineMap = append(l.lineMap, [2]int{begin, idx})
+			begin = idx + 1
+		}
+	}
+	// last line
+	l.lineMap = append(l.lineMap, [2]int{begin, idx + 1})
+	// fmt.Printf("l.lineMap %v\n", l.lineMap)
+}
+
+// CurrentPosition returns l.position
 func (l *Lexer) CurrentPosition() int {
 	return l.position
 }
 
+// LinePosition (pos) returns lineNum, begin, end
+func (l *Lexer) LinePosition(pos int) (int, int, int) {
+	idx := 0
+	begin := 0
+	end := 0
+	for i, tuple := range l.lineMap {
+		idx = i
+		begin, end = tuple[0], tuple[1]
+		if pos >= begin && pos <= end {
+			break
+		}
+	}
+	lineNum := idx + 1
+	return lineNum, begin, end
+}
+
+// GetLinePos (pos=l.position) returns lineNum, column, thisLine
+// optional arg: pos=l.position i.e. l.CurrentPosition()
+func (l *Lexer) GetLinePos(args ...int) (int, int, string) {
```

Comment:
I'd like to keep naming a bit more consistent -- the previous function was called `LinePosition` and this is called `GetLinePos`.

Could we maybe settle on `LinePosition()` and `ErrorPosition`?